core/types/transaction.go

```diff
@@ -488,6 +503,47 @@ func (t *TransactionsByPriceAndNonce) CurrentSize() int {
 	return len(t.heads)
 }
 
+//Forward move t to be one index behind tx, tx cant be nil
+func (t *TransactionsByPriceAndNonce) Forward(tx *Transaction) {
+	if tx == nil {
+		txTmp := t.Peek()
```

Comment:
This do not make too much sense. The tx can be nil only when 1. The `Loop` inside `commitTransactions` does not start yet. 2. The `Loop` ends.  Fo the \#1 case, you need to nothong, for the \#2 case, we should abort prefethcer.

Reply:
1. `prefetcher` start before the `woker` routine execute `*tx`, yet i've set some check process that would avoid the *start* nil case in the `prefetcher`, i think we shouldn't rely on the caller to skip the nil check in the method itself.
2. as for the *end* case, `worker` routine would break and then set the signal(close the channel) to abort `prefetcher`, there seems no guarantee that the concurrent `prefecher` routine would catch that signal immediately before touch the nil variable, yet we shouldn't rely on the caller as mentioned before.

Reply:
\#2, I think the `shift` operation is CPU intensive, break and wait for the nil variable is more CPU friendly.

Reply:
updated